tools/src/git_ops.rs

```diff
@@ -43,3 +43,164 @@ pub fn get_blame_lines(git_data: Option<&GitData>, blame_commit: &Option<Commit>
         _ => None,
     }
 }
+
+#[derive(Debug)]
+pub struct LineMap {
+    // Stores pairs (a, b) sorted so that `a` is always increasing.
+    // `a` is a line number, and `b` is the amount of adjustment needed
+    // for line numbers from `a` onwards. Subsequent entries effectively
+    // override previous ones, rather than being cumulative.
+    adjustments: Vec<(u32, i32)>,
+}
+
+impl LineMap {
+    pub fn map_line(&self, lineno: u32) -> u32 {
+        // Find the entry (a, b) in `adjustments` with the largest
+        // a <= lineno. This will give us the adjustment needed for
+        // lineno
+        let mut line_shift = 0;
+        for (ref start, ref shift) in &self.adjustments {
```

Comment:
You're right, binary searching would probably help some here. Even with the caching this would help because we do this lookup multiple times in a file (for different lines in that file). I might do that in a followup though, since I'll want to make sure to test it properly as well.